finesse/hardware/plugins/decades/decades_sensors.py

```diff
@@ -0,0 +1,128 @@
+"""This module provides an interface to the DECADES API.
+
+This is used to query the DECADES server for aircraft sensor data.
+"""
+import json
+import time
+
+from PySide6.QtCore import QTimer, QUrlQuery, Slot
+from PySide6.QtNetwork import QNetworkReply
+
+from finesse.config import (
+    DECADES_HOST,
+    DECADES_SENSORS_POLL_INTERVAL,
+    DECADES_SENSORS_QUERY_LIST,
+    DECADES_SENSORS_TOPIC,
+    DECADES_SENSORS_URL,
+)
+from finesse.hardware.device import Device
+from finesse.hardware.http_requester import HTTPRequester
+
+
+def get_decades_sensor_data(content: dict[str, list]) -> dict[str, float]:
+    """Parse and return sensor data from a DECADES server query.
+
+    Args:
+        content: The content of the HTTP response from the DECADES server
+
+    Returns:
+        None
+    """
+    parsed_content = {}
+    for sensor, value in content.items():
+        if len(value) != 1:
+            # Only single values (current time) should be returned
+            raise DecadesError(f"Unexpected number of values for sensor {sensor}")
+
+        parsed_content[sensor] = value[0]
+
+    return parsed_content
+
+
+@Slot()
+def _on_reply_received(reply: QNetworkReply) -> dict[str, float]:
+    """Handle received HTTP reply.
+
+    Args:
+        reply: the response from the server
+    """
+    if reply.error() != QNetworkReply.NetworkError.NoError:
+        raise DecadesError(f"Error: {reply.errorString()}")
+
+    content = json.loads(reply.readAll().data().decode())
+    return get_decades_sensor_data(content)
+
+
+class DecadesError(Exception):
+    """Indicates that an error occurred while querying the DECADES server."""
+
+
+class DecadesSensorsBase(
+    Device,
+    name=DECADES_SENSORS_TOPIC,
+    description="DECADES sensors",
+):
+    """An interface for monitoring generic DECADES sensor servers."""
+
+    def __init__(self, url: str) -> None:
+        """Create a new DECADES sensor monitor.
+
+        Args:
+            url: Address of the DECADES sensor data.
+        """
+        super().__init__()
+        self._url: str = url
+        self._requester = HTTPRequester()
+
+        # Poll device once on open.
+        # TODO: Run this synchronously so we can check that things work before the
+        # device.opened message is sent
+        self.send_data()
+
+    def send_data(self) -> None:
+        """Request the sensor data from the DECADES server.
+
+        The HTTP request is made on a background thread.
```

Comment:
Maybe it's some of the !T magic, but where is it set that the request is done in a background thread?

Reply:
Not actually sure. This part was copied from `em27_sensors.py`. Should probably have checked,